File: src/lib/validation.ts
import type { BreathSession, SoundSettings } from "@/lib/types";

const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

export function isUuid(value: string) {
  return UUID_PATTERN.test(value);
}

function isIntegerBetween(value: unknown, min: number, max: number) {
  return Number.isInteger(value) && Number(value) >= min && Number(value) <= max;
}

export function isValidSession(value: unknown): value is BreathSession {
  if (!value || typeof value !== "object") return false;
  const session = value as Partial<BreathSession>;
  if (typeof session.id !== "string" || !isUuid(session.id)) return false;
  if (session.status !== "completed" && session.status !== "stopped") return false;
  if (session.pace !== "slow" && session.pace !== "normal" && session.pace !== "fast") return false;
  if (!isIntegerBetween(session.plannedRounds, 1, 8) || !isIntegerBetween(session.breathsPerRound, 10, 60)) return false;
  if (typeof session.startedAt !== "string" || typeof session.completedAt !== "string") return false;
  const startedAt = Date.parse(session.startedAt);
  const completedAt = Date.parse(session.completedAt);
  if (!Number.isFinite(startedAt) || !Number.isFinite(completedAt) || completedAt < startedAt) return false;
  if (!Array.isArray(session.rounds) || session.rounds.length > 8) return false;
  return session.rounds.every((round) =>
    isIntegerBetween(round.roundIndex, 1, 8) &&
    isIntegerBetween(round.breathsCompleted, 10, 60) &&
    isIntegerBetween(round.retentionSeconds, 1, 3600),
  );
}

export function isValidSettings(value: unknown): value is SoundSettings {
  if (!value || typeof value !== "object") return false;
  const settings = value as Partial<SoundSettings>;
  return (
    (settings.musicTrack === "bambou" || settings.musicTrack === "meditation" || settings.musicTrack === "serenite") &&
    isIntegerBetween(settings.musicVolume, 0, 100) &&
    isIntegerBetween(settings.breathVolume, 0, 100) &&
    typeof settings.hapticsEnabled === "boolean"
  );
}

File: native/Sources/EoleCore/Validation.swift
import Foundation

// Validation des données persistées et importées.

private func makeISOFormatter(fractional: Bool) -> ISO8601DateFormatter {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = fractional
        ? [.withInternetDateTime, .withFractionalSeconds]
        : [.withInternetDateTime]
    return formatter
}

public func isUuid(_ value: String) -> Bool {
    UUID(uuidString: value) != nil
}

private func isIntegerBetween(_ value: Int, min: Int, max: Int) -> Bool {
    value >= min && value <= max
}

func parseDate(_ value: String) -> Date? {
    makeISOFormatter(fractional: true).date(from: value)
        ?? makeISOFormatter(fractional: false).date(from: value)
}

public func isValidSession(_ session: BreathSession) -> Bool {
    guard isUuid(session.id) else { return false }
    guard isIntegerBetween(session.plannedRounds, min: 1, max: 8) else { return false }
    guard isIntegerBetween(session.breathsPerRound, min: 10, max: 60) else { return false }
    guard let startedAt = parseDate(session.startedAt),
          let completedAt = parseDate(session.completedAt),
          completedAt >= startedAt
    else { return false }
    guard session.rounds.count <= 8 else { return false }
    return session.rounds.allSatisfy { round in
        isIntegerBetween(round.roundIndex, min: 1, max: 8)
            && isIntegerBetween(round.breathsCompleted, min: 10, max: 60)
            && isIntegerBetween(round.retentionSeconds, min: 1, max: 3600)
    }
}

public func isValidSettings(_ settings: SoundSettings) -> Bool {
    isIntegerBetween(settings.musicVolume, min: 0, max: 100)
        && isIntegerBetween(settings.breathVolume, min: 0, max: 100)
}
